vault/logical_system_paths.go

```diff
@@ -270,13 +270,17 @@ func (b *SystemBackend) sealPaths() []*framework.Path {
 
 func (b *SystemBackend) pluginsCatalogPath() *framework.Path {
 	return &framework.Path{
-		Pattern: "plugins/catalog/(?P<name>.+)",
+		Pattern: "plugins/catalog/(?P<type>auth|database|secret)/(?P<name>.+)",
```

Comment:
Since this endpoint has changed, we should update the dev flag/`addPlugin` to properly handle this: https://github.com/hashicorp/vault/blob/run-all-buitins-as-plugins/command/server.go#L985-L991

Reply:
```
› vault server -dev -dev-root-token-id=root -dev-plugin-dir=/Users/cleung/code/tmp/vault_plugins -log-level trace
==> Vault server configuration:

             Api Address: http://127.0.0.1:8200
                     Cgo: disabled
         Cluster Address: https://127.0.0.1:8201
              Listener 1: tcp (addr: "127.0.0.1:8200", cluster address: "127.0.0.1:8201", max_request_duration: "1m30s", max_request_size: "33554432", tls: "disabled")
               Log Level: trace
                   Mlock: supported: false, enabled: false
                 Storage: inmem
                 Version: Vault v1.0.0-beta1
             Version Sha: f48eb7a228aa0c25a00c7e235ebe50a681b2d888+CHANGES

Error enabling plugin my-mock-plugin: 1 error occurred:
        * unsupported path
```

Reply:
I ended up adding a flag to denote the plugin type on a directory: https://github.com/hashicorp/vault/pull/5536/commits/0d6c304004572096ff302c421a40bb4c34cc0f21 and https://github.com/hashicorp/vault/pull/5536/commits/c5f2e81213d7d8eedc7ebfe94c3cbf9f622d5f05

Reply:
What if the directory contains plugins for more than one type?

Reply:
In the current situation, it could only contain one so you'd maybe want 3 different directories for auth/secret/database plugins. However, I could change this to work differently. Like for instance, I could look for a string of auth, secret, or database in the plugin binary's name, and categorize it to match. Let me know what you think the best approach would be, I'm open to other ideas as well.

Reply:
Maybe we can load all the plugins the "old" way (in the base path of catalogView) and then call Brian's upgrade logic to have that automatically upgrade to the correct sub-view.

Reply:
Just pushed this! Take a look at https://github.com/hashicorp/vault/pull/5536/commits/0931816df96d42325410a0b008350476d1672edf